pkg/plugins/document/dynamodb/dynamodb.go

```diff
@@ -285,69 +322,91 @@ func (s *DynamoDocService) Query(collection *document.Collection, expressions []
 		)
 	}
 
-	queryResult := &document.QueryResult{
-		Documents: make([]document.Document, 0),
+	if res, err := s.query(collection, expressions, limit, pagingToken); err != nil {
+		return nil, newErr(
+			codes.Internal,
+			"query error",
+			err,
+		)
+	} else {
+		return res, nil
 	}
+}
 
-	// If partition key defined then perform a query
-	if collection.Parent != nil && collection.Parent.Id != "" {
-		err := s.performQuery(collection, expressions, limit, pagingToken, queryResult)
-		if err != nil {
+func (s *DynamoDocService) QueryStream(collection *document.Collection, expressions []document.QueryExpression, limit int) document.DocumentIterator {
+	newErr := errors.ErrorsWithScope(
+		"DynamoDocService.QueryStream",
+		map[string]interface{}{
+			"collection": collection,
+		},
+	)
+
+	colErr := document.ValidateQueryCollection(collection)
+	expErr := document.ValidateExpressions(expressions)
+
+	if colErr != nil || expErr != nil {
+		// Return an error only iterator
+		return func() (*document.Document, error) {
 			return nil, newErr(
-				codes.Internal,
-				"query error",
-				err,
+				codes.InvalidArgument,
+				"invalid arguments",
+				fmt.Errorf("collection error:%v, expression error: %v", colErr, expErr),
 			)
 		}
+	}
 
-		remainingLimit := limit - len(queryResult.Documents)
+	var tmpLimit = limit
+	var documents []document.Document
+	var pagingToken map[string]string
 
-		// If more results available, perform additional queries
-		for remainingLimit > 0 &&
-			(queryResult.PagingToken != nil && len(queryResult.PagingToken) > 0) {
+	// Initial fetch
+	res, fetchErr := s.query(collection, expressions, limit, nil)
```

Comment:
`s.query` is an eager query, with the full limit provided I think this will retrieve all the pages at once. The eager portion of `query` should probably be moved into `Query` so the code can be reused like this.

Reply:
done